src/firestarter/LoadWorker.cpp

```diff
@@ -47,39 +48,41 @@
 
 namespace firestarter {
 
-void Firestarter::initLoadWorkers() {
-  Environment->setCpuAffinity(0);
+void Firestarter::initLoadWorkers(const ThreadAffinity& Affinity) {
+  // Bind this thread to the first available CPU.
+  Topology->bindCallerToOsIndex(*Affinity.CpuBind.cbegin());
 
   // setup load variable to execute low or high load once the threads switch to
   // work.
   LoadVar = Cfg.Load == std::chrono::microseconds::zero() ? LoadThreadWorkType::LoadLow : LoadThreadWorkType::LoadHigh;
 
-  auto NumThreads = Environment->requestedNumThreads();
-
   // create a std::vector<std::shared_ptr<>> of requestenNumThreads()
   // communication pointers and add these to the threaddata
   if (Cfg.ErrorDetection) {
-    for (uint64_t I = 0; I < NumThreads; I++) {
+    for (uint64_t I = 0; I < Affinity.RequestedNumThreads; I++) {
       auto* CommPtr = static_cast<uint64_t*>(AlignedAlloc::malloc(2 * sizeof(uint64_t)));
       assert(CommPtr);
       ErrorCommunication.emplace_back(std::shared_ptr<uint64_t>(CommPtr, AlignedAlloc::free));
-      log::debug() << "Threads " << (I + NumThreads - 1) % NumThreads << " and " << I << " commPtr = 0x"
-                   << std::setfill('0') << std::setw(sizeof(uint64_t) * 2)
+      log::debug() << "Threads " << (I + Affinity.RequestedNumThreads - 1) % Affinity.RequestedNumThreads << " and "
+                   << I << " commPtr = 0x" << std::setfill('0') << std::setw(sizeof(uint64_t) * 2)
                    << std::hex
                    // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
                    << reinterpret_cast<uint64_t>(CommPtr);
     }
   }
 
-  for (uint64_t I = 0; I < NumThreads; I++) {
-    auto Td = std::make_shared<LoadWorkerData>(I, std::cref(*Environment), std::ref(LoadVar), Cfg.Period,
-                                               Cfg.DumpRegisters, Cfg.ErrorDetection);
+  auto CpuBindIt = Affinity.CpuBind.cbegin();
+  for (uint64_t I = 0; I < Affinity.RequestedNumThreads; I++, CpuBindIt++) {
```

Comment:
Since Affinity.CpuBind has the length Affinity.RequestedNumThreads (according to a comment in the header), do we really need to do it this way? imo, we should change that to `for (auto CpuBindIt : Affinity.CpuBind)`. If necessary, define I outside of the loop and increase it within.

Reply:
I'll change it to that. Ideally we would want to use `enumerate` but we will have to wait some more years for C++23 support. https://en.cppreference.com/w/cpp/ranges/enumerate_view